### PR TITLE
[ctx_prof] Automatically convert available external linkage to local for modules with contextual roots

### PR DESCRIPTION
**Pull request by mtrofin**
For the modules containing context roots, the way IPO happens will potentially result in imported functions that are differently specialized (even if themselves not inlined) than their originals. So we want to convert them to local rather than elide them.

Eventually we'd perform this as a ThinLTO directive.

### CODE REVIEW

**Comment by mtrofin** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109203?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#109203** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109203?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#109185** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109184** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109184?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109183** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109183?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mtrofin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/lib/Transforms/IPO/ElimAvailExtern.cpp [hunk 1]
```diff
@@ -125,8 +126,10 @@ static bool eliminateAvailableExternally(Module &M) {
 }
 
 PreservedAnalyses
-EliminateAvailableExternallyPass::run(Module &M, ModuleAnalysisManager &) {
-  if (!eliminateAvailableExternally(M))
-    return PreservedAnalyses::all();
+EliminateAvailableExternallyPass::run(Module &M, ModuleAnalysisManager &MAM) {
+  auto *CtxProf = MAM.getCachedResult<CtxProfAnalysis>(M);
+  if (!eliminateAvailableExternally(M, (CtxProf && !!(*CtxProf))))
```
Comment (teresajohnson) [inline llvm/lib/Transforms/IPO/ElimAvailExtern.cpp:131]: add a comment about why

File: llvm/test/Transforms/EliminateAvailableExternally/transform-to-local.ll [hunk 2]
```diff
@@ -1,6 +1,11 @@
 ; REQUIRES: asserts
 ; RUN: opt -passes=elim-avail-extern -avail-extern-to-local -stats -S 2>&1 < %s | FileCheck %s
+; RUN: echo '[{"Guid":1234, "Counters": [1]}]' | llvm-ctxprof-util fromJSON --input=- --output=%t_profile.ctxprofdata
+; RUN: opt -passes='assign-guid,require<ctx-prof-analysis>,elim-avail-extern' -use-ctx-profile=%t_profile.ctxprofdata -stats -S 2>&1 < %s | FileCheck %s
```
Comment (teresajohnson) [inline llvm/test/Transforms/EliminateAvailableExternally/transform-to-local.ll:4]: comment

**Review by teresajohnson — APPROVED** [review]
lgtm with a couple of comment suggestions

File: llvm/lib/Transforms/IPO/ElimAvailExtern.cpp (hunk 1 above)
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/ElimAvailExtern.cpp]: done

File: llvm/test/Transforms/EliminateAvailableExternally/transform-to-local.ll (hunk 2 above)
Comment (mtrofin) [reply llvm/test/Transforms/EliminateAvailableExternally/transform-to-local.ll:8]: done

**Comment by mtrofin** [comment]
### Merge activity

* **Sep 24, 4:17 PM EDT**: @mtrofin started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/109203).

File: llvm/lib/Transforms/IPO/ElimAvailExtern.cpp [hunk 3]
```diff
@@ -125,8 +126,16 @@ static bool eliminateAvailableExternally(Module &M) {
...
+EliminateAvailableExternallyPass::run(Module &M, ModuleAnalysisManager &MAM) {
+  auto *CtxProf = MAM.getCachedResult<CtxProfAnalysis>(M);
+  // Convert to local instead of eliding if we use contextual profiling in this
+  // module. This is because the IPO decisions performed with contextual
+  // information will likely differ from decisions made without. For a function
+  // that's imported, its optimizations will, thus, differ, and be specialized
+  // for this contextual information. Eliding it in favor of the original would
+  // undo these optimizations.
+  if (!eliminateAvailableExternally(M, /*Convert=*/(CtxProf && !!(*CtxProf))))
+    ;
```
Comment (RKSimon) [inline llvm/lib/Transforms/IPO/ElimAvailExtern.cpp:138]: @mtrofin This looks suspicious (static analysis warned about the repeated returns below) -  bad semicolon?

**Comment by mikaelholmen** [comment]
If I compile with EXPENSIVE_CHECKS on a large number of lit tests fail with this patch.
E.g.
```build-all-expensive/bin/llvm-lit -av ../lld/test/wasm/lto/thin-archivecollision.ll```
fails with
```
LLVM ERROR: Function @blah changed by EliminateAvailableExternallyPass without invalidating analyses
```

File: llvm/lib/Transforms/IPO/ElimAvailExtern.cpp (hunk 3 above)
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/ElimAvailExtern.cpp:138]: yup. thanks for flagging. Fixed in 35ae7ee.

**Pull request merged**